From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] oleaut32: Copy record fields properly in GetField/PutField/PutFieldNoCopy.

IRecordInfo::GetField places a copy of the field in the variant (a new
BSTR, an AddRef'd interface, the VARIANT's value; it handed out the
record's own BSTR, and refused interface and VARIANT fields); PutField
coerces and copies the value and releases the field's old one (it leaked
it), also for INVOKE_PROPERTYPUTREF (it returned E_NOTIMPL); PutFieldNoCopy
stores a value of the field's type as it is, taking ownership (it returned
E_NOTIMPL).

Gate: test/oleaut32-recinfo-gate.sh (64/32).
---
 dlls/oleaut32/recinfo.c | 173 +++++++++++++++++++++++++++++-------------------
 1 file changed, 104 insertions(+), 69 deletions(-)

diff --git a/dlls/oleaut32/recinfo.c b/dlls/oleaut32/recinfo.c
index dfe7625..93f2873 100644
--- a/dlls/oleaut32/recinfo.c
+++ b/dlls/oleaut32/recinfo.c
@@ -55,46 +55,94 @@ static inline IRecordInfoImpl *impl_from_IRecordInfo(IRecordInfo *iface)
     return CONTAINING_RECORD(iface, IRecordInfoImpl, IRecordInfo_iface);
 }
 
+/* field types a VARIANT can hold by reference: VariantCopyInd copies them
+ * (a BSTR is duplicated, an interface AddRef'd, a VARIANT copied) */
+static BOOL field_is_variant_type(enum VARENUM vt)
+{
+    switch (vt)
+    {
+    case VT_I2: case VT_I4: case VT_R4: case VT_R8: case VT_CY: case VT_DATE: case VT_BSTR:
+    case VT_DISPATCH: case VT_ERROR: case VT_BOOL: case VT_VARIANT: case VT_UNKNOWN: case VT_DECIMAL:
+    case VT_I1: case VT_UI1: case VT_UI2: case VT_UI4: case VT_I8: case VT_UI8: case VT_INT: case VT_UINT:
+        return TRUE;
+    default:
+        return FALSE;
+    }
+}
+
+/* GetField: "places a copy of the field's value in the variant" */
 static HRESULT copy_to_variant(void *src, VARIANT *pvar, enum VARENUM vt)
 {
+    VARIANT ref;
+
     TRACE("%p %p %d\n", src, pvar, vt);
 
-#define CASE_COPY(x) \
-    case VT_ ## x: \
-        memcpy(&V_ ## x(pvar), src, sizeof(V_ ## x(pvar))); \
-        break 
-
-    switch(vt) {
-        CASE_COPY(I2);
-        CASE_COPY(I4);
-        CASE_COPY(R4);
-        CASE_COPY(R8);
-        CASE_COPY(CY);
-        CASE_COPY(DATE);
-        CASE_COPY(BSTR);
-        CASE_COPY(ERROR);
-        CASE_COPY(BOOL);
-        CASE_COPY(DECIMAL);
-        CASE_COPY(I1);
-        CASE_COPY(UI1);
-        CASE_COPY(UI2);
-        CASE_COPY(UI4);
-        CASE_COPY(I8);
-        CASE_COPY(UI8);
-        CASE_COPY(INT);
-        CASE_COPY(UINT);
-        CASE_COPY(INT_PTR);
-        CASE_COPY(UINT_PTR);
+#ifndef SG_MUTANT_RECINFO_SHALLOW
+    if (field_is_variant_type(vt))
+    {
+        V_VT(&ref) = VT_BYREF | vt;
+        V_BYREF(&ref) = src;
+        return VariantCopyInd(pvar, &ref);
+    }
+#else
+    (void)ref;
+#endif
+
+    switch (vt)
+    {
+    case VT_INT_PTR:
+    case VT_UINT_PTR:
+        memcpy(&V_INT_PTR(pvar), src, sizeof(V_INT_PTR(pvar)));
+        break;
+#ifdef SG_MUTANT_RECINFO_SHALLOW
+    case VT_I4: memcpy(&V_I4(pvar), src, 4); break;
+    case VT_BSTR: memcpy(&V_BSTR(pvar), src, sizeof(BSTR)); break;
+#endif
     default:
         FIXME("Not supported type: %d\n", vt);
         return E_NOTIMPL;
-    };
-#undef CASE_COPY
-
+    }
     V_VT(pvar) = vt;
     return S_OK;
 }
 
+/* releases what a field of these types holds */
+static void release_field(void *field, enum VARENUM vt)
+{
+    switch (vt)
+    {
+    case VT_BSTR:
+        SysFreeString(*(BSTR *)field);
+        *(BSTR *)field = NULL;
+        break;
+    case VT_UNKNOWN:
+    case VT_DISPATCH:
+        if (*(IUnknown **)field) IUnknown_Release(*(IUnknown **)field);
+        *(IUnknown **)field = NULL;
+        break;
+    case VT_VARIANT:
+        VariantClear(field);
+        break;
+    default:
+        break;
+    }
+}
+
+/* stores the value of a variant of the field's type (taking its contents) */
+static void store_field(VARIANT *var, void *dest, enum VARENUM vt)
+{
+    release_field(dest, vt);
+    if (vt == VT_VARIANT)
+        *(VARIANT *)dest = *var;
+    else if (vt == VT_DECIMAL)
+        memcpy(dest, &V_DECIMAL(var), sizeof(DECIMAL));
+    else
+        memcpy(dest, &V_UI1(var), vt == VT_I8 || vt == VT_UI8 || vt == VT_R8 || vt == VT_CY || vt == VT_DATE ? 8 :
+               vt == VT_I2 || vt == VT_UI2 || vt == VT_BOOL ? 2 : vt == VT_I1 || vt == VT_UI1 ? 1 :
+               vt == VT_BSTR || vt == VT_UNKNOWN || vt == VT_DISPATCH ? sizeof(void *) : 4);
+}
+
+/* PutField: the variant's value, coerced and copied, replaces the field's */
 static HRESULT copy_from_variant(VARIANT *src, void *dest, enum VARENUM vt)
 {
     VARIANT var;
@@ -102,41 +150,19 @@ static HRESULT copy_from_variant(VARIANT *src, void *dest, enum VARENUM vt)
 
     TRACE("(%p(%d) %p %d)\n", src, V_VT(src), dest, vt);
 
-    hres = VariantChangeType(&var, src, 0, vt);
-    if(FAILED(hres))
-        return hres;
-
-#define CASE_COPY(x) \
-    case VT_ ## x: \
-        memcpy(dest, &V_ ## x(&var), sizeof(V_ ## x(&var))); \
-        break
-
-    switch(vt) {
-        CASE_COPY(I2);
-        CASE_COPY(I4);
-        CASE_COPY(R4);
-        CASE_COPY(R8);
-        CASE_COPY(CY);
-        CASE_COPY(DATE);
-        CASE_COPY(BSTR);
-        CASE_COPY(ERROR);
-        CASE_COPY(BOOL);
-        CASE_COPY(DECIMAL);
-        CASE_COPY(I1);
-        CASE_COPY(UI1);
-        CASE_COPY(UI2);
-        CASE_COPY(UI4);
-        CASE_COPY(I8);
-        CASE_COPY(UI8);
-        CASE_COPY(INT);
-        CASE_COPY(UINT);
-        CASE_COPY(INT_PTR);
-        CASE_COPY(UINT_PTR);
-    default:
-        FIXME("Not supported type: %d\n", V_VT(&var));
+    if (!field_is_variant_type(vt))
+    {
+        FIXME("Not supported type: %d\n", vt);
         return E_NOTIMPL;
-    };
-#undef CASE_COPY
+    }
+    VariantInit(&var);
+    if (vt == VT_VARIANT)
+        hres = VariantCopyInd(&var, src);
+    else
+        hres = VariantChangeType(&var, src, 0, vt);
+    if (FAILED(hres))
+        return hres;
+    store_field(&var, dest, vt);
     return S_OK;
 }
 
@@ -445,10 +471,8 @@ static HRESULT WINAPI IRecordInfoImpl_PutField(IRecordInfo *iface, ULONG wFlags,
             || (wFlags != INVOKE_PROPERTYPUTREF && wFlags != INVOKE_PROPERTYPUT))
         return E_INVALIDARG;
 
-    if(wFlags == INVOKE_PROPERTYPUTREF) {
-        FIXME("wFlag == INVOKE_PROPERTYPUTREF not supported\n");
-        return E_NOTIMPL;
-    }
+    /* PUTREF "just assigns the value of the variant ... using normal
+     * coercion rules", as PUT does here */
 
     for(i=0; i<This->n_vars; i++)
         if(!wcscmp(This->fields[i].name, szFieldName))
@@ -466,7 +490,7 @@ static HRESULT WINAPI IRecordInfoImpl_PutFieldNoCopy(IRecordInfo *iface, ULONG w
     IRecordInfoImpl *This = impl_from_IRecordInfo(iface);
     int i;
 
-    FIXME("%p, %#lx, %p, %s, %p stub\n", iface, wFlags, pvData, debugstr_w(szFieldName), pvarField);
+    TRACE("%p, %#lx, %p, %s, %p.\n", iface, wFlags, pvData, debugstr_w(szFieldName), pvarField);
 
     if(!pvData || !szFieldName || !pvarField
             || (wFlags != INVOKE_PROPERTYPUTREF && wFlags != INVOKE_PROPERTYPUT))
@@ -478,7 +502,18 @@ static HRESULT WINAPI IRecordInfoImpl_PutFieldNoCopy(IRecordInfo *iface, ULONG w
     if(i == This->n_vars)
         return TYPE_E_FIELDNOTFOUND;
 
-    return E_NOTIMPL;
+    /* "Passes ownership of the data to the assigned field ... does not copy
+     * the data": a value of the field's type goes in as it is; another type
+     * is coerced into a copy */
+#ifndef SG_MUTANT_RECINFO_NOCOPY_COPIES
+    if (field_is_variant_type(This->fields[i].vt)
+            && (This->fields[i].vt == VT_VARIANT || V_VT(pvarField) == This->fields[i].vt))
+    {
+        store_field(pvarField, ((PBYTE)pvData)+This->fields[i].offset, This->fields[i].vt);
+        return S_OK;
+    }
+#endif
+    return copy_from_variant(pvarField, ((PBYTE)pvData)+This->fields[i].offset, This->fields[i].vt);
 }
 
 static HRESULT WINAPI IRecordInfoImpl_GetFieldNames(IRecordInfo *iface, ULONG *pcNames,
